app/src/androidTest/java/com/github/warnastrophy/core/ui/map/MapViewModelTest.kt

```diff
@@ -1,23 +1,253 @@
+   * the app from crashing or getting stuck in a loading state.
+   */
+  @Test
+  fun startLocationUpdates_handle_null_location() = runTest {
+    val mockClient = mock<FusedLocationProviderClient>()
+    val callbackCaptor = argumentCaptor<LocationCallback>()
+
+    // Mock requestLocationUpdates to capture callback
+    doAnswer {
+          val callback = it.getArgument<LocationCallback>(1)
+          val locationResult = LocationResult.create(listOf(null))
+          callback.onLocationResult(locationResult)
+          null
+        }
+        .whenever(mockClient)
+        .requestLocationUpdates(any(), callbackCaptor.capture(), anyOrNull())
+
+    viewModel.startLocationUpdates(mockClient)
+    advanceUntilIdle()
+
+    val state = viewModel.uiState.value
+    assertFalse(state.isLoading)
+    assertEquals("No location fix available", state.errorMsg)
+  }
+
+  /**
+   * Verifies that the ViewModel correctly handles a `SecurityException` when starting location
+   * updates.
+   *
+   * This test simulates a scenario where location permissions have not been granted. It mocks the
+   * `FusedLocationProviderClient` to throw a `SecurityException` when `requestLocationUpdates` is
+   * called. It then asserts that the ViewModel's UI state is updated to reflect this error,
+   * specifically by setting an appropriate error message and ensuring the loading state is false.
+   */
+  @Test
+  fun startLocationUpdates_need_permissions() = runTest {
+    val mockClient = mock<FusedLocationProviderClient>()
+    val callbackCaptor = argumentCaptor<LocationCallback>()
+
+    // Mock requestLocationUpdates to throw SecurityException
+    whenever(mockClient.requestLocationUpdates(any(), callbackCaptor.capture(), anyOrNull()))
+        .thenThrow(SecurityException())
+
+    viewModel.startLocationUpdates(mockClient)
+    advanceUntilIdle() // advances all running coroutines to completion
+
+    val state = viewModel.uiState.value
+    assertFalse(state.isLoading)
+    assertEquals("Location permission not granted !", state.errorMsg)
+  }
+
+  /**
+   * Tests the behavior of `startLocationUpdates` when the underlying `requestLocationUpdates` call
+   * from the `FusedLocationProviderClient` throws a generic `RuntimeException`.
+   *
+   * This test verifies that:
+   * 1. The `ViewModel` correctly catches the exception.
+   * 2. The UI state is updated to reflect that loading has finished (`isLoading` is false).
+   * 3. An appropriate error message, derived from the exception, is set in the UI state.
+   */
+  @Test
+  fun startLocationUpdates_throw_exception() = runTest {
+    val mockClient = mock<FusedLocationProviderClient>()
+    val callbackCaptor = argumentCaptor<LocationCallback>()
+    val e = RuntimeException("Something went wrong")
+
+    // Mock requestLocationUpdates to throw Exception
+    whenever(mockClient.requestLocationUpdates(any(), callbackCaptor.capture(), anyOrNull()))
+        .thenThrow(e)
```

Comment:
You don’t actually use the callback here, the exception is thrown before the callback is invoked. So you can simplify: 
```suggestion
    whenever(mockClient.requestLocationUpdates(any(), any(), anyOrNull()))
    .thenThrow(SecurityException())

```

Reply:
I will apply this change to the two tests